swap/htlc.go

```diff
@@ -512,3 +561,227 @@ func (h *HtlcScriptV2) MaxTimeoutWitnessSize() int {
+// controlBlock for a.
+func (h *HtlcScriptV3) genControlBlock(leafScript []byte) ([]byte, error) {
+	var outputKeyYIsOdd bool
+
+	// Check for odd bit
+	if h.TaprootKey[0] == secp.PubKeyFormatCompressedOdd {
+		outputKeyYIsOdd = true
+	}
+
+	// Generate proof with unused script path
+	leaf := txscript.NewBaseTapLeaf(leafScript)
+	proof := leaf.TapHash()
+
+	internalPubKey, err := btcec.ParsePubKey(h.InternalPubKey)
+	if err != nil {
+		return nil, err
+	}
+
+	controlBlock := txscript.ControlBlock{
+		InternalKey:     internalPubKey,
+		OutputKeyYIsOdd: outputKeyYIsOdd,
+		LeafVersion:     txscript.BaseLeafVersion,
+		InclusionProof:  proof[:],
+	}
+	controlBlockBytes, err := controlBlock.ToBytes()
+
+	if err != nil {
+		return nil, err
+	}
+	return controlBlockBytes, nil
+}
+
+// genSuccessWitness returns the success script to spend this htlc with
+// the preimage.
+func (h *HtlcScriptV3) genSuccessWitness(
+	receiverSig []byte, preimage lntypes.Preimage) (wire.TxWitness, error) {
+
+	controlBlockBytes, err := h.genControlBlock(h.TimeoutScript)
+	if err != nil {
+		return nil, err
+	}
+	return wire.TxWitness{
+		preimage[:],
+		receiverSig,
+		h.ClaimScript,
+		controlBlockBytes,
+	}, nil
+}
+
+// GenTimeoutWitness returns the timeout script to spend this htlc after
+// timeout.
+func (h *HtlcScriptV3) GenTimeoutWitness(
+	senderSig []byte) (wire.TxWitness, error) {
+
+	controlBlockBytes, err := h.genControlBlock(h.ClaimScript)
+	if err != nil {
+		return nil, err
+	}
+	return wire.TxWitness{
+		senderSig,
+		h.TimeoutScript,
+		controlBlockBytes,
+	}, nil
+}
+
+// IsSuccessWitness checks whether the given stack is valid for
+// redeeming the htlc.
+func (h *HtlcScriptV3) IsSuccessWitness(witness wire.TxWitness) bool {
+	return len(witness) == 4
+}
+
+// Script is not implemented, but necessary to conform to interface.
+func (h *HtlcScriptV3) Script() []byte {
+	return nil
+}
+
+// MaxSuccessWitnessSize returns the maximum witness size for the
+// success case witness.
+func (h *HtlcScriptV3) MaxSuccessWitnessSize() int {
+	// Calculate maximum success witness size
+	//
+	// - number_of_witness_elements: 1 byte
+	// - sigLength: 1 byte
+	// - sig: 64 bytes
+	// - preimage_length: 1 byte
+	// - preimage: 32 bytes
+	// - witness_script_length: 1 byte
+	// - witness_script: len(script) bytes
+	// - control_block_length: 1 byte
+	// - control_block: 4129 bytes
+	return 1 + 1 + 64 + 1 + 32 + 1 + len(h.ClaimScript) + 1 + 4129
+}
+
+// MaxTimeoutWitnessSize returns the maximum witness size for the
+// timeout case witness.
+func (h *HtlcScriptV3) MaxTimeoutWitnessSize() int {
+	// Calculate maximum timeout witness size
+	//
+	// - number_of_witness_elements: 1 byte
+	// - sigLength: 1 byte
+	// - sig: 64 bytes
+	// - witness_script_length: 1 byte
+	// - witness_script: len(script) bytes
+	// - control_block_length: 1 byte
+	// - control_block: 4129 bytes
+	return 1 + 1 + 64 + 1 + len(h.TimeoutScript) + 1 + 4129
```

Comment:
Similarly here, we should probably calculate the worst control block size for our use case (even though we don't use this function).